### PR TITLE
feat(behavior_path_planner): add dynamic obstacle avoidance module

### PR DESCRIPTION
**Pull request by takayuki5168**
## Description

**This feature is under development, and disabled by default.**
NOTE:
- This module is supposed to be used with a new behavior_path_planner architecture
- The following parameter should be false since the trajectory may be outside the drivable area in some cases.
https://github.com/autowarefoundation/autoware_launch/blob/5910b08f7742d78ac5f1ade182d470032e40e398/autoware_launch/config/planning/scenario_planning/lane_driving/motion_planning/obstacle_avoidance_planner/obstacle_avoidance_planner.param.yaml#L7
- This module does not require RTC's approval to execute.

Added a dynamic avoidance module for avoiding dynamic obstacles which are running on the road.
In the current implementation, by extracting the dynamic obstacles to avoid from the drivable area, the motion planner plans the trajectory to avoid the obstacles.

[Screencast from 2023年04月27日 17時49分10秒.webm](https://user-images.githubusercontent.com/20228327/234811180-b429dcee-e766-4993-80e8-6d155be138b1.webm)

TODO
- The trajectory shape ahead of the ego is a little bit jerky.
    - Needs to improve the motion planner which generates the trajectory
    - Needs to improve the lateral controller not to be so sensitive to the trajectory shape.
- The dynamic avoidance module has to be assured to launch with the avoidance or lane_change modules simultaneously.
- Complicated cases cannot be dealt with by the current implementation
    - e.g. The case where there are obstacles on the left and right side very close to the ego's trajectory.

small TODOs
- Decision-making to avoid or not can be improved
    - not to avoid obstacles that have not been avoided.
    - not to conflict with the decision from the cruise and stop planner.

## Related links

launcher: https://github.com/autowarefoundation/autoware_launch/pull/299

## Tests performed

In the planning simulator, in a simple case, it's working correctly.
Though the steering controller is a little bit unstable.

## Interface changes

Nothing

## Effects on system behavior

Dynamic avoidance feature can be enabled.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/scene_module/dynamic_avoidance/dynamic_avoidance_module.cpp [hunk 1]
```diff
@@ -0,0 +1,392 @@
+// Copyright 2021 Tier IV, Inc.
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/src/scene_module/dynamic_avoidance/dynamic_avoidance_module.cpp:1]:
```suggestion
// Copyright 2021 Tier IV, Inc.
```
```suggestion
// Copyright 2023 TIER IV, Inc.
```

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module/dynamic_avoidance/dynamic_avoidance_module.hpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
+// Copyright 2021 Tier IV, Inc.
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/include/behavior_path_planner/scene_module/dynamic_avoidance/dynamic_avoidance_module.hpp:1]:
```suggestion
// Copyright 2023 TIER IV, Inc.
```

File: planning/behavior_path_planner/docs/behavior_path_planner_dynamic_avoidance_design.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# Dynamic avoidance design
+
+## Purpose / Role
+
+This is a module designed for avoiding obstacles which are running.
+Static obstacles such as parked vehicles are dealt with by the avoidance module.
+
+This module is under development.
+In the current implementation, the dynamic obstacles to avoid is extracted from the drivable area.
+Then the motion planner, in detail obstacle_avoidance_planner, will generate an avoiging trajectory.
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/docs/behavior_path_planner_dynamic_avoidance_design.md:10]:
```suggestion
Then the motion planner, in detail obstacle_avoidance_planner, will generate an avoiding trajectory.
```

**Review by satoshi-ota — COMMENTED** [review]
Could you add config in `behavior_path_planner/config/~ `?

https://github.com/autowarefoundation/autoware_launch/pull/299

File: planning/behavior_path_planner/src/scene_module/dynamic_avoidance/dynamic_avoidance_module.cpp [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+  for (const auto & object : target_objects_) {
+    const auto obstacle_poly = calcDynamicObstaclePolygon(*reference_path, object);
+    if (obstacle_poly) {
+      obstacle_polys.push_back(obstacle_poly.value());
+    }
+  }
+
+  BehaviorModuleOutput output;
+  output.path = reference_path;
+  output.reference_path = reference_path;
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/src/scene_module/dynamic_avoidance/dynamic_avoidance_module.cpp:193]: I think it should use previous module's turn signal info.

File: planning/behavior_path_planner/src/scene_module/dynamic_avoidance/dynamic_avoidance_module.cpp [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+  std::vector<tier4_autoware_utils::Polygon2d> obstacle_polys;
+  for (const auto & object : target_objects_) {
+    const auto obstacle_poly = calcDynamicObstaclePolygon(*reference_path, object);
+    if (obstacle_poly) {
+      obstacle_polys.push_back(obstacle_poly.value());
+    }
+  }
+
+  BehaviorModuleOutput output;
+  output.path = reference_path;
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/src/scene_module/dynamic_avoidance/dynamic_avoidance_module.cpp:192]: should use previous module's output

Comment (takayuki5168) [reply planning/behavior_path_planner/src/scene_module/dynamic_avoidance/dynamic_avoidance_module.cpp:185]: DONE
https://github.com/autowarefoundation/autoware.universe/pull/3415/commits/c3ffd01951319c61b10c7f4534cfd620b0227851

File: planning/behavior_path_planner/src/scene_module/dynamic_avoidance/dynamic_avoidance_module.cpp (hunk 4 above)
Comment (takayuki5168) [reply planning/behavior_path_planner/src/scene_module/dynamic_avoidance/dynamic_avoidance_module.cpp:186]: DONE
https://github.com/autowarefoundation/autoware.universe/pull/3415/commits/c3ffd01951319c61b10c7f4534cfd620b0227851

File: planning/behavior_path_planner/src/scene_module/dynamic_avoidance/dynamic_avoidance_module.cpp [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/src/scene_module/dynamic_avoidance/dynamic_avoidance_module.cpp:181]:
```suggestion
  output.path = getPreviousModuleOutput().path;
```

**Review by TakaHoribe — APPROVED** [review]
LGTM!

File: planning/behavior_path_planner/src/scene_module/dynamic_avoidance/dynamic_avoidance_module.cpp (hunk 6 above)
Comment (takayuki5168) [reply planning/behavior_path_planner/src/scene_module/dynamic_avoidance/dynamic_avoidance_module.cpp]: https://github.com/autowarefoundation/autoware.universe/pull/3415/commits/fbb2e3674ce3189c5e2c36146dfd2d2f9a7fa2c3
DONE

**Review by satoshi-ota — APPROVED** [review]

**Pull request merged**